### ISSUE TITLE
Closer integration with OSM

### ISSUE THREAD

**Issue by HarelM** [IsraelHikingMap/Site#258]
I think a closer integration with OSM might prove worthy.
Here are my thoughts:
1. Add a login button in one of the sides.
2. This login button will open a window that will allow granting the site permission to the user tracks.
3. Once this is done we can use the user's OSM data for all kind of stuff including simple editing.
@zstadler @valleyofdawn - this is where I need your help, I'm not sure where to add this data/what functionality we want to add - keep in mind that I still want to keep this site as simple as possible.
The first thing that comes to mind it to add a top navigation bar, but I'm not sure I want to add this as it was in the site's old design and I think the current way is a lot cleaner.
My other thought was to add a section below or above the routes with the label "OSM tracks" or something of that sort which will allow the user to interact with them.
Bottom line here is that the user will be able to create a route on IHM and send it to OSM and back, which will allow us to consume it from mobile app in the future, maybe...

Any thought are welcome, I'm still brain storming.
This link allows a very simple OSM OAuth integration (managed to verify it is working and saw my tracks). 
https://github.com/osmlab/osm-auth
I have created an account for IHM:
מפתח צרכן: H5Us9nv9eDyFpKbBTiURf7ZqfdBArNddv10n6R6U
סוד צרכן: ccYaQUKLz26XEzbNd8uWoQ6HwbcnrUUp8milXnXG

**Comment by zstadler** [comment]
Suggestions for leveraging OSM authentication:
- [ ] Enhance the shared maps capability and allow registered users to manage their maps:
  - Add title and description to each shared map. This was also requested for social media sharing.
  - Show the user's shared maps with its info: name, description, creation date, no. of uses, last use date, last modified date.
  - Edit shared maps title and description.
  - Load a shared map from the list.
  - Save current map into an existing shared map.
- [ ] Add a layer of the user's traces on OSM. One possible UI is:
  - Show a POI at the starting point of each trace
  - POI text to include trace title, date and optional description
  - Click on the POI text will download the trace to the map. The user will be able to create a new layer from the trace, similar to handling of search results today.
- [ ] Allow simple editing of existing (only) OSM POIs.
    - Since new POIs can derived from the Satellite Imagery layer, I suggest we do not include this facility.
- [ ] Smart diff between an OSM trace and OSM map data used for OSM map update
  - Analyze a trace and find sections of it that are missing from the OSM map data (not the rendered map!)
  - Allow updating the map with the missing way(s). 
    - Each new way will be connected to existing roads, if any, at its start, end, and intersections. 
    - The user will provide some way info:
      - Mandatory: way type (track, path, etc.)
      - Optional: name, name:en, colour, route type, route operator, more(?)
      - Use colour and route info to add way to a compatible neighboring route relation, or create a new route relation.

**Comment by HarelM** [comment]
This list is very long and should be probably split into a few issue.
If not split into few issues it should be prioritized.
First glimpse into the integration:
![image](https://cloud.githubusercontent.com/assets/3269297/20011499/85e0a316-a2b4-11e6-8916-9ab367c5ce4a.png)
I can login and see my name and my photo.
I added it in the upper section of the layers sidebar only for now, if you think it should be in another place let me know.

**Comment by zstadler** [comment]
I would prefer the login button and its status to be always visible as part of "nudging" the community into OSM involvement.

Not so sure about the picture...

**Comment by HarelM** [comment]
The picture is not negotiable, I'm taking it from osm, you can leave it empty there.

Since the site does not have a top navigation bar (and I think it's better without it) I don't have a better place to put it.

Let's start this way, and I'll continue to think of a better place to put it.

**Comment by HarelM** [comment]
I needed another properties page, couldn't keep it in the layer menu - it was too complicated
New suggestion - move the languages down and add this on top.
![image](https://cloud.githubusercontent.com/assets/3269297/20029097/48339634-a34c-11e6-89d3-d49d66a341d8.png)
When clicking the wrench:
![image](https://cloud.githubusercontent.com/assets/3269297/20029108/7a666988-a34c-11e6-97d6-6f8f5b1e87ae.png)
What do you think?
Traces are the GPX files from OSM - clicking one of them will present it on the map much like the search results, but if it is selected and you click the edit in OSM it will place it in the background for editing, still need to think about the "finding missing parts" user scenario.
Shares are all the shares you did (This requires to add a name and a description to each share - which is long overdue anyway).
I'm still not sure what to do when you have a lot of them - either paging or search or both...
It can be a table, either with images or without.
@valleyofdawn @zstadler thoughts are very welcome at this state, I'm still thinking how it should be presented, later on I'll have something and it'll be harder for me to let go of what I did :-).

**Comment by HarelM** [comment]
@zstadler @valleyofdawn,
I'm trying out something I learned today - you don't need to develop anything in order to have the user have an initial feeling of a feature.
Have a look at the below presentation, you can click the upper right symbol in order to see how thing will look and feel - first click is for login (which will cause a popup of osm to authorize), next click will open the new menu.
[OsmIntegration.pptx](https://github.com/IsraelHikingMap/Site/files/573954/OsmIntegration.pptx)
Thoughts are welcome, feel free to alter this presentation in order to collaborate ideas.

**Comment by valleyofdawn** [comment]
Cool trick with the PPT.
I like the experience thermometer.
I find the traces not very useful. I personally stopped using them since
 can drag my GPX directly to the editor and since Strava gives a much
denser activity map than OSM traces.
Perhaps instead the editing heat-map I sent the other day? Provided the guy
agrees.
http://www.yosmhm.neis-one.org/
Or his stats from here:
http://hdyc.neis-one.org/

**Comment by zstadler** [comment]
I can definitely see the limited value of traces for iD users and/or bicycle riders. For me they have great value for mapping and for planning future trips. I find it hard to manage my traces on the desktop as 

I'm not familiar with a program that will efficiently show me their location on the map. It would be great of we can place the traces on the map. 

In any case, since we want to promote contributions to the map, I prefer map edits to come first.

**Comment by HarelM** [comment]
Got the PPT idea from a UX course I'm taking which makes you see things differently when it comes to users interacting with applications.
I think the thermometer is enough, of course the stats you linked have good data, but it's extremely verbose and too much information IMHO - I want to keep a clean and neat UI just to make the user want to map.
The tracks will be there as they are part of a bigger plan to engage the user in mapping, we will discuss it in the meeting on Wednesday.
I hope to upload a working version soon to the test site, although this change requires some changes in the database too, so I'm not sure I'll be able to push it in phases...
If you think something is missing in the screen let me know - as far as I can understand from this thread the screen is acceptable by all.

**Comment by valleyofdawn** [comment]
So there will be another way to add traces? We're aiming at new users. If
they have no easy way or reason to upload them, since iD doesn't support
that, or might be frustrating.

**Comment by HarelM** [comment]
I'm not sure I totally follow, traces are GPS created tracks, I think the best approach will be to upload them from a navigation app on your mobile device, not from IHM. 
Currently IHM discards all the timing data in the recording, so I'm not sure it's even possible without big changes to upload data to OSM traces.

**Comment by valleyofdawn** [comment]
So the answer is no.  Let's check and see how many of the users actually
upload traces to OSM. I have a few but they all predate iD. I'm saying
trace uploading is no longer a meaningful part of OSM contribution. If you
don't want to replace it with a useful state,  like a map or a graph of
mapping activity, just leave it blank.

**Comment by valleyofdawn** [comment]
Zeev, I'm using GPSies to manage GPX tracks. Many other options are out
there.

**Comment by HarelM** [comment]
I have uploaded an initial version with very limited capabilities - I know that this is not the final version.
http://israelhiking.osm.org.il/test_osm/
When user is not logged-in in the browser, it doesn't seem to work perfect, but I'm not sure there's anything I can do with it...

**Comment by zstadler** [comment]
QA results:
- User name is not updated after login on Firefox. Need to click "Unknown User" in order to refresh it.
- I don't see any of my private traces

**Comment by HarelM** [comment]
I don't have these issues:
No need to click after login (I was logged in to OSM though).
Can see my traces - I think they are public though.
![image](https://cloud.githubusercontent.com/assets/3269297/20120615/2c06135c-a616-11e6-93f9-84b7f34da3e9.png)

**Comment by HarelM** [comment]
My shares preview, planned site update tomorrow night.
![image](https://cloud.githubusercontent.com/assets/3269297/20122040/8458d9b6-a61d-11e6-870b-847201922a94.png)

**Comment by zstadler** [comment]
I have the same issues with Chrome and IE. 

It is enough to move the map in order for the username to be shown, but none of my traces are shown, including one public trace.

How do you want to proceed?

**Comment by zstadler** [comment]
Can it be because I don't have a profile picture?

**Comment by HarelM** [comment]
Could be, I'll try fixing it tonight, I did some changes to the database so I don't think I'll be able to load a test site.
Try committing out this line on the server and see if it starts working.

**Comment by HarelM** [comment]
@zstadler please let me know if you are now able to use OSM features.

**Comment by HarelM** [comment]
I'm not sure a description is actually needed, I think a title might be enough for 80% of the cases - most users won't know what to put in a description field...
Phase 3 requirements:
1. Allow user to delete a "share".
2. Allow user to see a trace on the map - current temporary behavior is seeing it in OSM.
3. Allow user to copy the share link with ease (copy to clipboard button).
4. Allow the user to add unmapped part to OSM - this is a big feature and might get a separate issue.
5. Solve security issue - users are not logged in on the server just on the client side which wasn't an issue until now - a user can "share" as other user, and delete other user's shares.
6. Allow user to update a share - title will be easy, not sure about current map state, mainly how to explain the user what he should do in order to achieve it.

**Comment by zstadler** [comment]
Trace list issues:
- Thumbnails of private traces are not shown because the OSM server returns a 403 Forbidden error.
- Link to a private trace on the OSM server receives a "Trace not found!" error. Perhaps the OSM authentication is not passed along with the link.
- The scroll-bar for the trace (and shares!?) list appears at the edge of the page rather than at the edge of the pop-up.
- A search box for the trace title, date, and filename would be a very nice addition.

I think we do need a description, even if the casual 80% of the uses (not necessarily users) will not use it. 
A non-empty description can also show in Facebook shares instead of our [boilerplate](https://en.wikipedia.org/wiki/Boilerplate_(text)):

> "בין אם אתם יוצאים לטיול רגלי, רכיבה על אופניים"...

**Comment by HarelM** [comment]
Can't reproduce issue with private route. I uploaded a private route and I can easily see it, including thumbnail.
Try logging out and logging in and see if all the check-box are checked:
![image](https://cloud.githubusercontent.com/assets/3269297/20238868/66b0fbd6-a8fd-11e6-8e07-0c4db9477cc9.png)

**Comment by zstadler** [comment]
Now I no longer see my private traces. Same results on Firefox and Chrome.

The ```gpx_files``` xml is received with all traces, but the network trace contains only the GET for the one public trace.

**Comment by HarelM** [comment]
I know, good spotting. Since I don't have a proper solution to the fact that the server can't fetch the private traces I decided to show just the public ones. I don't have a solution to this problem yet and I couldn't find one using Google.

**Comment by HarelM** [comment]
I think that most of the functionality is implemented. 
I'll start working on the "finding un-mapped parts".

**Comment by zstadler** [comment]
I still don't see my private traces uploaded to OSM, just the one public trace I have.
This issue happens on both Firefox and Chrome

**Comment by HarelM** [comment]
Can you explain the following? I can't understand what you wanted to implement here.

> Since new POIs can derived from the Satellite Imagery layer, I suggest we do not include this facility.

I'm also moving the finding missing routes to another issue, this issue is too long to find anything.
Bottom line from what I can understand the only thing that's missing here are the private traces, right?

**Comment by zstadler** [comment]
@HarelM wrote:
> Can you explain the following? I can't understand what you wanted to implement here.
>>    Since new POIs can derived from the Satellite Imagery layer, I suggest we do not include this facility.

In a nutshell, I suggest we do not implement a feature, although it is technically feasible.

The long story:
The comment is related to its parent proposal:
> - [ ] Allow simple editing of existing (only) OSM POIs.
>   - Since new POIs can derived from the Satellite Imagery layer, I suggest we do not include this facility.

Technically we can add "simple editing of existing (only) OSM POIs". However, I suggest we do not implement this proposal because of potential issues with basing an edit on a Satellite Imagery that was not approved as a source for OSM editing.

**Comment by HarelM** [comment]
I added private routes. 
@zstadler Please close this issue if you think there's nothing else to add that is not related to finding missing parts.

**Comment by zstadler** [comment]
Indeed, private traces are now shown.

Remaining issues:
- Please print the trace upload date and its tags. Note that the tag list may be empty.
- The grey scroll bar is hard to find on the side of the grayed-out map. the location is also misleading to think it applies to the map.

**Comment by HarelM** [comment]
* Not sure how useful the tag list is, if you can convince me it has a value for the users I'll add it, otherwise I won't.
 * I will add the date.
 * Not planning on fixing the scroll bar - it's too much work and this is the way bootstrap works - touch screen and scroll mouse button are your friends, side scroll bar is almost never used, especially in mobile devices where it is very very small...

**Comment by zstadler** [comment]
When opening a long trace list for the first time in the window, the message says for several seconds "No traces, you should really upload some to OSM".

<img width="338" alt="2016-11-30 08_46_24-clipboard" src="https://cloud.githubusercontent.com/assets/1304610/20742581/a417f8be-b6d9-11e6-9166-949923efe74f.png">

This is confusing and alarming. Please change the message to "Fetching your traces" and/or some "busy" animation.

**Comment by HarelM** [comment]
I don't have a long trace list, I can't fully test it, can you please check if now it's OK?
Please close this issue if so, I'm waiting on your input in order to check-in the code changes.

**Comment by zstadler** [comment]
I assume the difference is that the pop-up does not show until the list is downloaded.
This means I click the button and nothing happens, so I click again and still nothing happens...

It's hard to decide which is worse...

Previously, there was a text message that was replaced with a trace list once it arrived.
I suggest changing the wording of the initial message to "Fetching your traces...", and replacing it with a the "No traces, you should really upload some to OSM" message or the actual list when the results arrive.

**Comment by HarelM** [comment]
This is not what I wanted to achieve - I added a very nice looking loading animation so that the animation will show when you are waiting.
This is strange...
Will you agree to login with your user on the site on the server so I can test out things?

**Comment by HarelM** [comment]
It seems that the file was not sent to the server for some reason, in any case you can now try it out in the regular site (not in the test site).
I know you'll tell me that animation is not enough to let the user know that he need to wait ... :-)
If you'd like a different one, you can pick from the following site:
http://cssload.net/

**Comment by zstadler** [comment]
The spinner works and looks good!

Can you center-align it horizontally?

**Comment by HarelM** [comment]
Tried, there are issues with the hebrew when doing so, I think it's goo enough. Closing...
